### ISSUE TITLE
Deployment best practises for TorchServe

### ISSUE THREAD

**Issue by vishal-wiai** [pytorch/serve#668]
I have few questions regarding deployment using TorchServe

- How scalable is TorchServe? Both in horizontal as well as vertical scaling and possible tradeoffs?
- Which way is preferred for deployment on AWS : Sagemaker, EC2, ECS or something else? And Why?
- What's the difference in using TorchServe or Multi Model Server on Sagemaker? 
- Is it possible to do lambda deployment using TorchServe?
- Any possible way to do load testing?
- In what cases the `health` API returns "unhealthy" status?

**Comment by alvarobartt** [comment]
Hi! I'm also interested on the `curl http://localhost:8080/ping` behavior when returning either "Healthy" or "Unhealthy", as sometimes I got the "Unhealthy" response, but I didn't figure out what does it really mean... 😩 

A simple explanation over that would be really useful! 👍🏻 @dhaniram-kshirsagar @harshbafna

**Comment by msaroufim** [comment]
Hi @vishal-wiai 

Apologies for the delayed answer but some answers to your questions

> How scalable is TorchServe? Both in horizontal as well as vertical scaling and possible tradeoffs?
Currently it scales in the number of processes and workers that you assign to it, so scales horizontally very well. Main constraint is not supporting threading on a single process which is coming soon with our C++ backend work

> Which way is preferred for deployment on AWS : Sagemaker, EC2, ECS or something else? And Why?
I personally use EC2. Sagemaker is great if you want to deploy something out of a notebook so if you don't have much ops background it's the easiest. ECS is useful when you're using some larger orchestration framework like Kubernetes

> What's the difference in using TorchServe or Multi Model Server on Sagemaker?
TorchServe is the successor Multi Model Server, the READMEs even look very similar

> Is it possible to do lambda deployment using TorchServe?
Yep but I haven't personally done it https://aws.amazon.com/blogs/machine-learning/using-container-images-to-run-pytorch-models-in-aws-lambda/

> Any possible way to do load testing?
https://github.com/pytorch/serve/tree/master/benchmarks

> In what cases the health API returns "unhealthy" status?
Basically if the number of workers available is greater than 0 but number of workers allocated is 0 https://github.com/pytorch/serve/blob/master/frontend/server/src/main/java/org/pytorch/serve/util/ApiUtils.java#L312
